tensorstore/kvstore/tiled_tiff/tiled_tiff_key_value_store.cc

```diff
@@ -0,0 +1,389 @@
+#include <tiffio.h>
```

Comment:
This is a good starting point. 

Generally you will want to make it work with an underlying kvstore, so that means it will need a read/write cache. Look at tensorstore/kvstore/neuroglancer_uint64_sharded/neuroglancer_uint64_sharded.cc as a model.

I'd probably start with List, and add metadata manipulation to internal/image/tiff_{encoder,decoder}

Reply:
I am a little confused here. The usecase that motivated us to have a OmeTiff driver is to enable chuck based access. Usually these tiff files are big and at a given time, the user is interested in a small chunk of data. The decoder in internal/image/tiff_decoder seems to be reading the whole file while we want to read only a portion of the file. So, should not we have a separate driver and kvstore? According to the OME spec, these files are always 5D from an API point of view, however, each 2D slices are stored in a separate IFD inside the tiff file. The OmeXml metadata helps us to figure out the IFD value for a given Z, Channel and Time value. And then we can extract the (x_range, y_range) values using the ReadTileFromFile call.

Reply:
It appears that this ometiff driver just accesses the local filesystem directly, which means it would only work if the ometiff dataset is stored on the local filesystem --- instead it should use a base kvstore so that it works with any storage system.

The tensorstore kvstore interface allows you to specify a byte range for read requests to avoid reading the entire file.

Reply:
I also agree that the `ometiff` driver is tightly coupled with tiled `tiled_tiff` kvstore. Here is my difficulty in uncoupling them. 

I want to read a tile as a chunk and that makes me dependent on high-level API like `TiffReadTile(...)`. Also, when I am extracting the OmeXML metadata, I am using `TiffGetField` to read the IMAGE_DESCRIPTION tag. If I use a base kvstore, how should I approach it so that I do not have to borrow the implementation of these functions?

Reply:
> This is a good starting point.
> 
> Generally you will want to make it work with an underlying kvstore, so that means it will need a read/write cache. Look at tensorstore/kvstore/neuroglancer_uint64_sharded/neuroglancer_uint64_sharded.cc as a model.
> 
> I'd probably start with List, and add metadata manipulation to internal/image/tiff_{encoder,decoder}

So, if I understand it correctly, `ometiff` driver will still work with a single specific key-value store driver but the users will not be able to choose it via JSON spec, rather it will be sort of hard-coded in `ometiff` similar to `neuroglancer_uint64_sharded`?

Reply:
The use of tiled_tiff as an implementation detail is fine --- that should not be exposed to users.  But tiled_tiff should use a base kvstore (that is exposed to users) rather than just using the local filesystem.

Reply:
Got it. Thanks. I am looking at the neuroglacner driver to see that is done and will try to mimic similar design.

Reply:
Hello again, I spent some time to understand the neuroglancer_uint64_unsharded and did not actually get too far. 
Here are some of my questions related to `'ometiff' driver. These may appear very naive questions but I would really appreciate your guidance on them:

1. If the user is allowed to use 'file', 'memory', 'gcs' and 'http' as  key-value store driver and I want to still leverage `libtiff` API to extract metadata and various information from the file and also to read tiles, will it be possible use the `tiled_tiff` key-value storage driver (with some rearrangement)? Or will I need to reimplement most of the `libtiff` functionality here. I am not sure how to how to successfully use the byte-range read functionality of tensorstore interface without bringing in those `libtiff` implementation details (since they actually figures out which part of the file to read)

2. I see that `ShardedKeyValueStore` inherits from `internal_kvstore` and in the neuroglancer tensorstore driver, `GetDataKeyValueStore` is overrided. Is that the path you are suggesting me to take when you mean that `tiled_tiff` should use a base kvstore that is exposed to the users?

Reply:
> Hello again, I spent some time to understand the neuroglancer_uint64_unsharded and did not actually get too far. 
> Here are some of my questions related to `'ometiff' driver. These may appear very naive questions but I would really appreciate your guidance on them:
> 
> 1. If the user is allowed to use 'file', 'memory', 'gcs' and 'http' as  key-value store driver and I want to still leverage `libtiff` API to extract metadata and various information from the file and also to read tiles, will it be possible use the `tiled_tiff` key-value storage driver (with some rearrangement)? Or will I need to reimplement most of the `libtiff` functionality here. I am not sure how to how to successfully use the byte-range read functionality of tensorstore interface without bringing in those `libtiff` implementation details (since they actually figures out which part of the file to read)

Libtiff has its own file i/o abstraction, so you could potentially implement the interface libtiff expects on top of a kvstore, support to how the existing non-tiled tiff driver implements it on top of absl::Cord.  The difficulty is that libtiff uses a synchronous API, so that would only allow a single concurrent read operation at a time.  Working around that limitation might require either using some libtiff internals directly, or somehow cloning libtiff handles.

> 
> 2. I see that `ShardedKeyValueStore` inherits from `internal_kvstore` and in the neuroglancer tensorstore driver, `GetDataKeyValueStore` is overrided. Is that the path you are suggesting me to take when you mean that `tiled_tiff` should use a base kvstore that is exposed to the users?

Yes, the use of a base kvstore exposed to users should be similar to how it is done with ShardedKeyValueStore.  But note that GetDataKeyValueStore is part of the kvs_backed_chunk_driver abstraction, which I don't think is a good fit for tiled_tiff.  Instead tiled_tiff should probably use ChunkCache directly.